core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,287 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.exception.RetryableException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.http.HttpResponseHandler;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.internal.http.TransformingAsyncResponseHandler;
+import software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler;
+import software.amazon.awssdk.core.internal.http.async.SimpleHttpContentPublisher;
+import software.amazon.awssdk.core.internal.http.loader.DefaultSdkAsyncHttpClientBuilder;
+import software.amazon.awssdk.core.retry.RetryPolicyContext;
+import software.amazon.awssdk.http.HttpStatusFamily;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
+import software.amazon.awssdk.imds.Ec2MetadataAsyncClient;
+import software.amazon.awssdk.imds.Ec2MetadataRetryPolicy;
+import software.amazon.awssdk.imds.EndpointMode;
+import software.amazon.awssdk.imds.MetadataResponse;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+@Immutable
+@ThreadSafe
+public final class DefaultEc2MetadataAsyncClient implements Ec2MetadataAsyncClient {
+
+    private static final Logger log = Logger.loggerFor(DefaultEc2MetadataClient.class);
+    private static final int DEFAULT_RETRY_THREAD_POOL_SIZE = 3;
+
+    private final Ec2MetadataRetryPolicy retryPolicy;
+    private final SdkAsyncHttpClient httpClient;
+    private final EndpointMode endpointMode;
+    private final URI endpoint;
+    private final RequestMarshaller requestMarshaller;
+    private final Duration tokenTtl;
+    private final ScheduledExecutorService asyncRetryScheduler;
+
+    private DefaultEc2MetadataAsyncClient(Ec2MetadataAsyncBuilder builder) {
+        this.retryPolicy = builder.retryPolicy != null ? builder.retryPolicy
+                                                       : Ec2MetadataRetryPolicy.builder().build();
+        this.endpointMode = builder.endpointMode != null ? builder.endpointMode
+                                                         : DEFAULT_ENDPOINT_PROVIDER.resolveEndpointMode();
+        this.endpoint = builder.endpoint != null ? builder.endpoint
```

Comment:
Endpoint and endpoint mode should be mutually exclusive, right? Can we add validation? We can use this method https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/Validate.java#L786

Reply:
from the SEP:

> If endpoint is specified, the client MUST use the explicitly provided endpoint.
> 
> If endpoint_mode is supported and endpoint is unspecified
> 
> If set to IPv4, use http://169.254.169.254
> If set to IPv6, use http://[fd00:ec2::254]

By that, I understand that both `endpoint_mode` and `endpoint` could be specified, in which case we ignore `endpoint_mode` and use the `endpoint` value directly. What do you think?

Reply:
After discussion, we will implement them being mutually exclusive